Backoff feature

filip-lozanovski opened the pull request:
The PR is related to this JIRA ticket https://issues.corp.twilio.com/browse/DSYNC-1126.
Fixed a bug so the retrier does not schedule a new retry attempt when the maximum attempt time is reached.
The Backoff class that is added to the Retrier class is copied in functionality from
https://github.com/MathieuTurcotte/node-backoff and translated to Typescript. 
The purpose of this is to replace the Backoff exponential retrier in the Sync JS SDK
with the operational retrier. The code also has a bugfix for a bug in the original node-backoff
module around the randomization factor.

filip-lozanovski committed:
Change the yml file to use Gulp for building in Travis CI.

filip-lozanovski committed:
Added tslint file.
Added gulpfile with tasks for cleanup, linting, running unit tests and building the library.
Updated dependencies and added new packages in package.json needed for the new gulp tasks.
Fixed a bug so the retrier does not schedule a new retry attempt when the maximu8m attempt time is reached.
Added the Backoff class to the retrier with the same functionality as the open source backoff NPM module
and added test coverage for the backoff functionality.

filip-lozanovski committed:
Increment the patch version of the library.

filip-lozanovski committed:
Add gulpfile and tslint config to npmignore

dsineln commented on package.json:
```diff
@@ -11,6 +11,7 @@
   "author": "Sergei Chertkov",
   "license": "ISC",
   "devDependencies": {
+    "@babel/core": "^7.4.4",
```
unused dependencies in this list, which are safe to remove:
* @types/chai
* @types/chai-as-promised
* @types/mocha
* @types/node
* @types/sinon
* @types/sinon-as-promised
* @types/sinon-chai
* babelify
* jsdoc
* sinon-as-promised
* ts-node

filip-lozanovski replied:
how was this determined ?
i am absolutely sure some of these dependencies are necessary

dsineln replied:
depcheck utility shows this list if you run it against this repo. Also, wiping these dependencies does not yield any error in building of the lib or running tests.

dsineln commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "operation-retrier",
-  "version": "2.0.0",
+  "version": "2.0.1",
```
We fork original code (-> change source of publication), modify build scripts and move in whole new functionality of Backoff into the existing module. I'd say it deserves at least a minor version bump.

dsineln commented on src/retrier.ts:
```diff
@@ -136,33 +137,137 @@ class Retrier extends EventEmitter {
+  private readonly maxDelay: number;
+  private readonly initialDelay: number;
+  private readonly factor: number;
+  private randomisationFactor: number;
```
readonly

dsineln commented on src/retrier.ts:
```diff
@@ -25,14 +25,14 @@ class Retrier extends EventEmitter {
+  }) {
     super();
 
     this.minDelay = options.min;
```
all internal fields of Retrier which are assigned from options can be made readonly.

dsineln commented on src/retrier.ts:
```diff
@@ -100,6 +100,7 @@ class Retrier extends EventEmitter {
       this.cleanup();
       this.emit('failed', new Error('Maximum attempt time limit reached'));
       this.reject(new Error('Maximum attempt time limit reached'));
+      return;
```
Can we have a test on the actual fix?

dsineln commented on src/retrier.ts:
```diff
@@ -136,33 +137,137 @@ class Retrier extends EventEmitter {
+    options = options || {};
+
+    if (this.isDef(options.initialDelay) && options.initialDelay < 1) {
+      throw new Error('The initial timeout must be greater than 0.');
```
timeout -> delay
that 0. -> than 1

dsineln commented on src/retrier.ts:
```diff
@@ -136,33 +137,137 @@ class Retrier extends EventEmitter {
+    if (this.isDef(options.initialDelay) && options.initialDelay < 1) {
+      throw new Error('The initial timeout must be greater than 0.');
+    } else if (this.isDef(options.maxDelay) && options.maxDelay < 1) {
+      throw new Error('The maximal timeout must be greater than 0.');
```
maximal timeout -> max delay
than 0. -> than 1

dsineln commented on src/retrier.ts:
```diff
@@ -136,33 +137,137 @@ class Retrier extends EventEmitter {
+
+  public backoff(err?: any) {
+    if (this.timeoutID !== null) {
+      throw new Error('Backoff in progress.');
```
Do we need this error here? All we do is throw a error (aka return from the method). All we do in Sync JS SDK is `catch(e) {}` which is also waste of time and computation. If it's not used in other libs in any other way (aka we do not have proper error handling) I'd suggest simply doing a silent return and removing an exception handler in Sync JS SDK.

dsineln commented on src/retrier.ts:
```diff
@@ -136,33 +137,137 @@ class Retrier extends EventEmitter {
+    if (this.maxDelay <= this.initialDelay) {
+      throw new Error('The maximal backoff delay must be greater than the initial backoff delay.');
+    }
+    this.backoffDelay = 0;
```
How about to call `reset()` here instead? More robust, less code, code re-use? Also, remove `this.timeoutID = null;` above then.

dsineln commented on src/retrier.ts:
```diff
@@ -136,33 +137,137 @@ class Retrier extends EventEmitter {
+    this.nextBackoffDelay = this.initialDelay;
+
+    this.backoffNumber = 0;
+    clearTimeout(this.timeoutID);
```
I think we need a check for null/undefined before clearing it.

filip-lozanovski replied:
Null check is not needed, clearTimeout does not explode on  bad input

dsineln commented on src/retrier.ts:
```diff
@@ -136,33 +137,137 @@ class Retrier extends EventEmitter {
+    this.backoffNumber++;
+  }
 
+  private isDef(value) {
```
return type is missing

dsineln commented on src/retrier.ts:
```diff
@@ -136,33 +137,137 @@ class Retrier extends EventEmitter {
+    this.maxNumberOfRetry = -1;
+    this.backoffNumber = 0;
+
+    if (options && options.factor !== undefined) {
```
Let's have 'factor' validation similar to other options' fields validation, using is `ifDef` function.

dsineln commented on src/retrier.ts:
```diff
@@ -136,33 +137,137 @@ class Retrier extends EventEmitter {
+    this.backoffDelay = Math.min(this.nextBackoffDelay, this.maxDelay);
+    this.nextBackoffDelay = this.backoffDelay * this.factor;
+    let randomisationMultiple = 1 + Math.random() * this.randomisationFactor;
+    return Math.min(this.backoffDelay, Math.round(this.backoffDelay * randomisationMultiple));
```
This statement neglects whole point of randomization, since `Math.min(value, value * [1.0 ... 1.2])` is always going to be `value` because randomization factor makes second argument >= value all the time. So in order to fix bug "we can go over maxDelay" i'd suggest to change this statement to `Math.min(randomizedValue, this.maxDelay)`. This will keep delay growing exponentially with randomization factor till it exceeds maxDelay - at this point we stop increasing or randomizing it and will start returning maxDelay for any further backoff attempt.

dsineln commented on test/test.ts:
```diff
@@ -73,96 +84,275 @@ describe('Retrier', () => {
+    it('the ready event should be passed backoff number and the backoff delay', () => {
+      backoff.backoff();
+      mockClock.tick(10);
+      expect(backoffCallback).to.have.been.calledWith(0, 10);
```
Should be `readyCallback` here

filip-lozanovski replied:
good find

dsineln commented on test/test.ts:
```diff
@@ -73,96 +84,275 @@ describe('Retrier', () => {
+
+      expect(failCallback).not.to.have.been.called;
+      backoff.backoff(error);
+      expect(failCallback).to.have.been.calledWith(error);
```
Let's make sure backoffCallback has not been called here, to protect ourselves from some missing "return" in future.

dsineln commented on test/test.ts:
```diff
@@ -73,96 +84,275 @@ describe('Retrier', () => {
+      expect(failCallback).to.have.been.calledWith(error);
+    });
+
+    it('calling backoff while a backoff is in progress should throw an error', () => {
```
Test is potentially useless if we remove error here (as per some discussion above)

dsineln commented on test/test.ts:
```diff
@@ -73,96 +84,275 @@ describe('Retrier', () => {
+    it('the randomisation factor should be between 0 and 1', () => {
+      testFunc({randomisationFactor: 1.1}, 'The randomisation factor must be between 0 and 1.');
+      testFunc({randomisationFactor: -0.1}, 'The randomisation factor must be between 0 and 1.');
+      testFunc({randomisationFactor: 0}, 'The randomisation factor must be between 0 and 1.');
```
false positive here and below - we allow 0 and 1, so should not expect any error. And there is actually no error, and we successfully pass through `testFunc` because it is written in a wrong way (see comment above)

dsineln commented on test/test.ts:
```diff
@@ -73,96 +84,275 @@ describe('Retrier', () => {
+    function testFunc(initOptions, expectedMesage) {
+      try {
+        backoff = Backoff.exponential(initOptions);
+      } catch (err) {
```
Allows false-positives, if `Backoff.exponential` throws no errors we will never end up in catch-block and silently exit.

dsineln commented on test/test.ts:
```diff
@@ -73,96 +84,275 @@ describe('Retrier', () => {
+      });
+      for (let i = 0; i < 10; i++) {
+        let next = backoff.next();
+        expect(backoff.next()).to.be.below(601);
```
How this thing tests randomization? In this test we expect value _on each step_ to be somewhere between initialDelay and maxDelay, which is always true. So this test is also a false-positive - it does not test that we increase and/or randomize the `next()` step.

dsineln commented on test/test.ts:
```diff
@@ -73,96 +84,275 @@ describe('Retrier', () => {
+    it('delays should follow an exponential sequence', () => {
+      const expectedDelays = [10, 20, 40, 80, 160, 320, 640, 1000, 1000];
+      for (let delay of expectedDelays) {
+        expect(backoff.next()).to.equal(delay);
```
Currently we have both randomization broken AND backoff in tests is initialized with empty randFactor which makes it 0 by default. Pls make sure that initializing backoff with non-zero randFactor will fail this test (manually, to make sure randomization works).

dsineln commented on test/test.ts:
```diff
@@ -73,96 +84,275 @@ describe('Retrier', () => {
+      }
+    });
+
+    it('delays should restart from the initial delay after reset', () => {
```
i'd also suggest adding a test, that backoff is fully functional after reset (this is the main pattern how it's being used in Subscriptions engine):
 - go though couple of ticks, ensure `readyCallback` is called on each tick
 - do reset, clean test callback expectations
 - go again through couple of ticks, making sure that ticks are starting with initial delay and that `readyCallback` is properly called anew

filip-lozanovski committed:
Bumped the minor instead of the patch version.
Added readonly modifier where possible.
Added a test to verify the retrier does not go into an infinite retry loop after the maximum retry time is exceeded.

dsineln commented on test/test.ts:
```diff
@@ -142,6 +142,34 @@ describe('Retrier', () => {
+    myspy.onCall(0).returns(Promise.reject({code: 503, message: 'Server unavailable'}));
+    myspy.onCall(1).returns(Promise.reject({code: 503, message: 'Server unavailable'}));
+    myspy.onCall(2).returns(Promise.reject({code: 503, message: 'Server unavailable'}));
+    // myspy.onCall(4).returns(Promise.reject({code: 503, message: 'Server unavailable'}));
```
Pls wipe unnecessary comments with next commit (here and below)

filip-lozanovski committed:
Removed unused modules.
Fixed error messages, removed duplicate code, fixed tne randomized delays
and refactored the 'backoff' method to not throw an error when a backoff is already in progress.
Unit test fixes and improvements as required in the PR.

filip-lozanovski committed:
Modified a test to verify that the randomizer works properly.

Commit:
Cosmetic changes: added back devDependency since IDE syntax check is complaining; fixed typos; aligned limits for input options

dsineln commented:
+1
